README.md
# AutoML_Jupyter

CHANGELOG.md
# Change Log

All notable changes to the "auto-jupyter" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

server.py
import nbformat
from notebook.notebookapp import list_running_servers
import requests
import json

# Launch your Jupyter server first.
server_info = next(list_running_servers())

# Parse the .ipynb file.
with open("new_test.ipynb") as f:
    nb = nbformat.read(f, as_version=4)

# Identify the cell to execute.
cell_to_execute = nb.cells[0]  # for example

# Execute the code in the cell.
session_url = server_info['url'] + 'api/sessions'
headers = {'Authorization': 'token ' + server_info['token']}
session = json.loads(requests.get(session_url, headers=headers).text)[0]
requests.post(
    server_info['url'] + 'api/sessions/' + session['id'] + 'execute',
    headers=headers,
    data=json.dumps({
        'code': cell_to_execute.source,
        'silent': False,
        'store_history': True,
        'user_expressions': {},
        'allow_stdin': False
    })
)


server3.py
import nbformat
from nbconvert.preprocessors.execute import ExecutePreprocessor
from nbclient.exceptions import CellExecutionError, DeadKernelError
import copy



notebook_filename = 'new_test.ipynb'
notebook_filename_out = 'executed_new_test.ipynb'

# Load the notebook
with open(notebook_filename) as f:
    nb = nbformat.read(f, as_version=4)



resources = {'metadata': {'path': './'}}
ep = ExecutePreprocessor(timeout=600, kernel_name='python3')

def refine_cell(error_msg: str, original_cell: str) -> str:
    prompt = f"I would like to fix the following code. Please just give me the fixed version of this code, without any explanations or comments. \n {original_cell} \n for which I get the following error: \n {error_msg}"
    print("PROMPT: ", prompt)
    return "print(\"fixed\")"

for index, cell in enumerate(nb.cells):
    if cell.cell_type == "code":
        try:
            copied_nb = copy.deepcopy(nb)
            # Remove cells after cell i from the copied notebook
            copied_nb['cells'] = copied_nb['cells'][:index+1]

            out, resources = ep.preprocess(copied_nb, resources)
        except Exception as e:
            # print(e)
            nb.cells[index].source = refine_cell(str(e), nb.cells[index].source)
            



with open(f"notebooks/{notebook_filename_out}", mode='w', encoding='utf-8') as f:
    nbformat.write(nb, f)

print("finished execution")

src/api.ts
import axios from 'axios';
import { initialWrite } from './jupyter';

function preprocess(input: string): string {
  return input;
}

function postprocess(result: any): any {
  return result;
}

export async function generateCode(input: string, endpoint: string): Promise<void> {
  try {
    // Preprocess the input
    const preprocessedInput = preprocess(input);

    // Post the preprocessed input to the endpoint
    const response = await axios.post(endpoint, { input: preprocessedInput });

    // Get the result from the response
    const result = response.data.result;

    // Postprocess the result
    const postprocessedResult = postprocess(result);

    // Call the initialWrite function with the output
    initialWrite(postprocessedResult);
  } catch (error) {
    console.error('Error:', error);
  }
}

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { jupyter } from './jupyter';

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "auto-jupyter" is now active!');
	console.log('=============================================================');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('auto-jupyter.helloWorld', () => {
		vscode.window.showInformationMessage('Hello World from Auto_Jupyter!');
	}); 

	context.subscriptions.push(disposable);

	let disposable2 = vscode.commands.registerCommand(
		'auto-jupyter.readJupyter', 
		jupyter
	); 
	context.subscriptions.push(disposable2);

	// context.subscriptions.push(
	// 	vscode.workspace.onDidOpenTextDocument(readFirstMarkdownCell),
	// 	vscode.workspace.onDidSaveTextDocument(readFirstMarkdownCell),
	// 	vscode.workspace.onDidCloseTextDocument(readFirstMarkdownCell)
	// );


}

// This method is called when your extension is deactivated
export function deactivate() {}


FastAPI-Backend/app/main.py
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from google.auth import credentials
from google.oauth2 import service_account
import google.cloud.aiplatform as aiplatform
from vertexai.preview.language_models import CodeGenerationModel
from fastapi.openapi.docs import get_swagger_ui_html, get_redoc_html
import vertexai
import json 
import os
import openai
from pydantic import BaseModel
from typing import Optional, List

# Load the service account json file
# Update the values in the json file with your own
with open(
    "service_account.json"
) as f:  # replace 'serviceAccount.json' with the path to your file if necessary
    service_account_info = json.load(f)

with open("openai.json") as f:
    openai_json = json.load(f)
    print("openai_json: ", openai_json)
    openai.api_key = openai_json["api_key"]

my_credentials = service_account.Credentials.from_service_account_info(
    service_account_info
)

# Initialize Google AI Platform with project details and credentials
aiplatform.init(
    credentials=my_credentials,
)

with open("service_account.json", encoding="utf-8") as f:
    project_json = json.load(f)
    project_id = project_json["project_id"]


# Initialize Vertex AI with project and location
vertexai.init(project=project_id, location="us-central1")

# Initialize the FastAPI application
app = FastAPI()

# Configure CORS for the application
origins = ["http://localhost", "http://localhost:8080", "http://localhost:3000"]
origin_regex = r"https://(.*\.)?alexsystems\.ai"
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_origin_regex=origin_regex,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


@app.get("/")
async def root():
    """Root endpoint that returns available endpoints in the application"""
    return {
        "Endpoints": {
            "code": "/code",
        }
    }


@app.get("/docs")
async def get_documentation():
    """Endpoint to serve Swagger UI for API documentation"""
    return get_swagger_ui_html(openapi_url="/openapi.json", title="docs")


@app.get("/redoc")
async def get_documentation():
    """Endpoint to serve ReDoc for API documentation"""
    return get_redoc_html(openapi_url="/openapi.json", title="redoc")


class Item(BaseModel):
    prompt: str
    temperature: float = 0.5
    data: Optional[List[str]]

@app.post("/code")
async def generate_code(item: Item):
    """
    Endpoint to handle code generation.
    Receives a message from the user, processes it, and returns a response from the model.
    """
    temperature = item.temperature
    prompt = item.prompt
    parameters = {
        "temperature": temperature,  # Temperature controls the degree of randomness in token selection.
        "max_output_tokens": 2048,  # Token limit determines the maximum amount of text output.
    }

    dataPrompt = ""
    if (item.data is not None):
        data = item.data
        allData = ''.join(data)
        if "X =" in allData and "y =" in allData:
            exec(allData, globals())
            dataPrompt = f"The code should be compatible with input shape {X.shape[1:]} and output shape {y.shape[1:]}."


    content = f"Please write me pytorch code to {prompt}.Please only give code, and no comments, examples, or explanations. {dataPrompt}"

    print("content: ", content)
    # Call the model
    response = llm_inference(parameters, content)

    # Return the model's response
    return {"response": response}

@app.post("/summarize")
async def summarize_code(item: Item):
    """
    Endpoint to handle code summarization.
    Receives a message from the user, processes it, and returns a response from the model.
    """
    temperature = item.temperature
    prompt = item.prompt
    parameters = {
        "temperature": temperature,  # Temperature controls the degree of randomness in token selection.
        "max_output_tokens": 100,  # Token limit determines the maximum amount of text output.
    }
    content = f"Please write a markdown title for pytorch code in {prompt}."

    # Call the model
    response = llm_inference(parameters, content)

    # Return the model's response
    return {"response": response}

def llm_google(parameters, prompt):
    code_generation_model = CodeGenerationModel.from_pretrained("code-bison@001")
    response = code_generation_model.predict(
        prefix=prefix, **parameters
    )
    return response.text

def llm_inference(parameters, content):
    response = openai.ChatCompletion.create(
        # model="gpt-4",
        model="gpt-3.5-turbo",
        messages=[
            {
            "role": "user",
            "content": content,
            }
        ],
        temperature=parameters["temperature"],
        max_tokens=1024,
        top_p=1,
        frequency_penalty=0,
        presence_penalty=0
    )
    return response


src/jupyter.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import { ServerConnection, Kernel, ServiceManager } from '@jupyterlab/services';
import { spawn } from 'child_process';
import { URL } from 'url';
import axios, { AxiosResponse } from 'axios';

const genCodeEndpoint = 'http://localhost:8080/code';
const sumCodeEndpoint = 'http://localhost:8080/summarize';

export async function jupyter(): Promise<void> {

  // Make sure a text editor is active
  if (!vscode.window.activeTextEditor) {
    console.log("No active text editor found");
    return;
  }

  const document = vscode.window.activeTextEditor?.document;
  const editorPath = vscode.window.activeTextEditor.document.uri.fsPath;
  // const editorDir = vscode.workspace.getWorkspaceFolder(document.uri);
  const editorDir = path.dirname(editorPath);
  const notebook: any = parseDocumentToJson(document!);

  console.log("Printing notebook:");
  console.log(notebook);

  // console.log("Server connection");
  // // Start a new Jupyter server
  // let jupyter = await spawn('jupyter', ['notebook', '--no-browser'], { cwd: editorDir });

  // let baseUrl: string | undefined;
  // let token: any;
  // // Handle output - for some weird reason, it runs on stderr
  // jupyter.stderr.on('data', data => {
  //   let output = data.toString();
  //   console.log('INIT SEQUENCE:', output);

  //   // Parse output for URL and token
  //   let urlMatch = output.match(/http:\/\/localhost:\d+\/\?token=\w+/);
  //   if (urlMatch) {
  //     let url = new URL(urlMatch[0]);
  //     baseUrl = url.origin;
  //     token = url.searchParams.get('token');
  //     console.log('INIT SEQUENCE - Jupyter server base URL:', baseUrl);
  //     console.log('INIT SEQUENCE - Jupyter server token:', token);

  //     // Call the openAndExecuteCell function after getting the token
  //     // openAndExecuteCell(baseUrl, token, notebook, editorPath);
  //   }
  // });

  // // Handle stdout
  // jupyter.stdout.on('data', data => {
  //   console.log('Jupyter Server threw stout:', data.toString());
  // });

  // // Check if the process exited normally
  // jupyter.on('exit', code => {
  //   console.log("process exited")
  //   if (code !== 0) {
  //     console.log(`Jupyter Eito process exited with code ${code}`);
  //   }
  // });

  // console.log("Server connection");
  // // Create a server settings object
  // let serverSettings: ServerConnection.ISettings = ServerConnection.makeSettings({
  //   baseUrl: baseUrl,
  //   token: token,
  // });

  // console.log("Server connection object created");
  // // Create a service manager
  // let manager: ServiceManager.IManager = new ServiceManager({ serverSettings });

  // console.log("Created manager object");

  // // Start a new kernel
  // let model: Kernel.IModel;
  // manager.kernelspecs.ready.then(() => {
  //   manager.kernels.startNew({ name: 'python3' })  // Change 'python3' to the name of your kernel
  //     .then(kernel => {
  //       model = kernel.model;
  //       console.log('Kernel started:', kernel.id);
  //     })
  //     .catch(err => {
  //       console.error('Failed to start kernel:', err);
  //     });
  // });

  // console.log("Kernel started");

  /*
  === NOTEBOOK STRUCTURE ===
  */

  // Find the first markdown cell that begins with the string "!auto-jupyter" and if none found, return
  const firstMarkdownCell = notebook.cells.find((cell: any) => {
    return cell.cell_type === 'markdown' && cell.source[0].trim().startsWith('!auto-jupyter');
  });
  if (!firstMarkdownCell) {
    console.log("No markdown cell found");
    return;
  }

  console.log("Found markdown cell with !auto-jupyter, executing...");

  // Get the code from the code cell before the markdown cell
  const data = notebook.cells[notebook.cells.indexOf(firstMarkdownCell) - 1].source;

  // Get the text after the "!auto-jupyter" string
  const command = firstMarkdownCell.source[0].trim().split(' ').slice(1).join(' ');
  
  // Post the command to the endpoint and catch errors
  const temperature = 0.5;
  let codeOutput: any;
  let codeSummary: any;
  try {
    console.log(`Posting command to endpoint... Command: ${command}`);
    codeOutput = (await axios.post(genCodeEndpoint, {prompt: command, data: data, temperature: temperature})).data.response.choices[0].message.content;
    console.log(`Code output: ${codeOutput}`);
    codeSummary = (await axios.post(sumCodeEndpoint, {prompt: codeOutput, temperature: temperature})).data.response.choices[0].message.content;
    console.log(`Code summary: ${codeSummary}`);
  } catch (error: any) {
    console.error("Error posting command to endpoint");
    if (error.response) {
      console.error("Response data:", error.response.data);
      console.error("Response status:", error.response.status);
    } else if (error.request) {
      console.error("Request data:", error.request);
    } else {
      console.error("Other error:", error.message);
    }
    return;
  }

  console.log("Successfully posted command to endpoint, writing to file...");
  console.log("=============================================================");
  // TODO: Check if there is empty code
  // TODO: Delete !auto-jupyter line
  // Insert the body of the result into the code cell
  notebook.cells[notebook.cells.indexOf(firstMarkdownCell) + 1].source = codeOutput;

  // Replace the markdown cell text with the codeSummary
  notebook.cells[notebook.cells.indexOf(firstMarkdownCell)].source = codeSummary;

  const newFileContent = JSON.stringify(notebook, null, 2);
  fs.writeFileSync(document.uri.fsPath, newFileContent, 'utf8');
}


const parseDocumentToJson = (document: vscode.TextDocument) => {
  // Make sure a document is open
  if (!document) {
    console.log("No active document found");
    return;
  }

  // Read the entire file content
  let fileContent: string;
  try {
    fileContent = fs.readFileSync(document.uri.fsPath, 'utf8');
  } catch (err) {
    console.error(`Error reading file: ${err}`);
    return;
  }

  // Parse the content as JSON
  let notebook: any;
  try {
    notebook = JSON.parse(fileContent);
  } catch (err) {
    console.error(`Error parsing JSON: ${err}`);
    return;
  }
  return notebook;
}


// async function openAndExecuteCell(baseUrl: string, token: string | null, notebook: any, editorPath: string): Promise<void> {
//   try {

//     // 1. Create a new session
//     console.log('CREATE SESSION - Creating session...');
//     let response = await axios.post(
//       `${baseUrl}/api/sessions`,
//       {
//         notebook: { path: editorPath }, // replace with your notebook file path
//         kernel: { name: 'python3' } // replace with your kernel name
//       },
//       {
//         headers: { Authorization: `token ${token}` },
//       }
//     );
//     console.log('CREATE SESSION - Session created:', response.data);

//     const sessionId = response.data.id;

//     // 2. Execute code in the newly created session
//     const code = 'print("Hello, World!")'; // replace with the code of the cell you want to execute

//     response = await axios.post(
//       `${baseUrl}/api/sessions/${sessionId}/execute`,
//       { code },
//       {
//         headers: { Authorization: `token ${token}` },
//       }
//     );

//     console.log('EXECUTE CODE - Code executed:', response.data);

//   } catch (error: any) {
//       console.error('Error executing cell:', error);
//       console.error(error.arg1.stack);
//   }
// }
